doc/compiler/optimizer/BenefitInliner.md

```diff
@@ -0,0 +1,94 @@
+# BenefitInliner
+
+BenefitInliner introduces a novel way of inlining. Compared to the traditional way of inlining relies on hard-coded heuristics, BenefitInliner provides a systematic way of evaluating inlining candidates that combines the dynamic and static benefits of inlining. 
+
+## Overview of main phases of BenefitInliner
+1. Build the search space for candidate methods to be inlined.
+    1. **IDTBuilder** constructs an inlining dependency tree (**IDT**).
+    2. **AbstractInterpreter** generates an **InliningMethodSummary** stashing potential optimization opportunities and specifying the constraints for parameters to make optimizations happen.
+    3. **AbstractInterpreter** computes abstract state values for each program state. 
+    4. At each call site, we calculate the benefit of inlining this method with its call frequency (direct benefit) and optimizations unlocked (indirect benefit) according to **InliningMethodSummary**.
+2. Run the knapsack algorithm for choosing the best inlining plan.  
+3. Commit the inlining plan and inline the methods.
+    
+## Higher-Level Structure of the parts of BenfitInliner
+
+### 1. BenefitInliner.cpp
+
+Class hierarchy of the adapter class:
+* **TR::Optimization**
+    * **BenefitInlinerWrapper**
+
+Class hierarchy of the real inliner:
+* **TR_InlinerBase**
+    * **BenefitInlinerBase**
+       * **BenefitInliner**
+
+* **BenefitInlinerWrapper** is an adapter class which extends from **TR::Optimization**. `BenefitInlinerWrapper::perform()` initializes a **BenefitInliner** and controls how benefit inlining process is being done.
+
+* **BenefitInliner** is the real Inliner that contains functionalities for selecting inlining candidates. When a **BenefitInliner** is initialized, it comes with an inlining budget that constrains inlining by avoiding an uncontrollable increase in binary size. The inlining budget will be calculated based on the hotness and size of the method being JIT compiled. `BenefitInliner::buildingInliningDependecyTree()` builds an **IDT** that contains all the candidate methods to be inlined. `BenefitInliner::inlinerPacking()` consumes the **IDT** and selects the optimal set of methods to be inlined given the inlining budget and the cost and benefit of each method in the **IDT**. 
+
+* **BenefitInlinerBase** contains the functionalities of doing the actual inlining according to the inlining proposal proposed by `BenefitInliner::inlinerPacking()`. 
+
+### 2. InliningProposal.cpp
+
+* **InliningProposal** proposes an optimal set of methods that will be inlined. **BenefitInlinerBase** takes **InliningProposal** as a dependency. **InliningProposal** will be set by the `BenefitInliner::inlinerPacking()` and will be consulted during the inlining. 
+
+* **InliningProposalTable** is a 2D table of **InliningProposal**. By the nature of `BenefitInliner::inlinerPacking()` being a nested knapsack algorithm, a 2D table is required. 
```

Comment:
I thought `InliningProposal` _was_ the table? What exactly is the distinction between `InliningProposal` and `InliningProposalTable`?

Reply:
The InliningProposalTable is a 2D table of InliningProposal. InliningProposal encapsulates a bit vector which indicates which method should be inlined and which should not. 

By looking at the traditional knapsack algorithm, which selects the optimal set of goods which has the maximum values but does not overweight the weight limits. The algorithm uses a 2D table for dynamic programming computation, an element in the table is a selection plan, which corresponds to the inliningProposal in this case.